Kernel panic (arithmetic overflow) in InitStackWriter::write_bytes during execve with large environment

### Describe the bug

The kernel crashes with a panic (attempt to subtract with overflow) when handling an execve syscall that passes an extremely large number of environment variables. The panic occurs in kernel/src/process/process_vm/init_stack/mod.rs at line 381, within the InitStackWriter logic.

# https://github.com/asterinas/asterinas/blob/8ace76978eadd74dd2952866c269ac88b41c577b/kernel/src/process/process_vm/init_stack/mod.rs#L381

### To Reproduce

1. Compile the program and run.

```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <string.h>

/*
 * PoC for potential arithmetic overflow in adjust_stack_alignment
 * 
 * The crash occurs in the kernel when calculating auxvec_size:
 * let auxvec_size = (self.auxvec.table().len() + 1) * (size_of::<u64>() * 2);
 * 
 * This can overflow if the auxiliary vector becomes too large.
 * The auxiliary vector size can be influenced by:
 * 1. Large number of environment variables
 * 2. Very long program arguments
 * 3. Specific binary characteristics that require many aux entries
 * 
 * This PoC attempts to trigger the overflow by creating a process with
 * an extremely large number of environment variables, which should
 * result in a large auxiliary vector during execve.
 */

#define NUM_ENV_VARS 1000000

int main() {
    pid_t pid;
    char **envp;
    int i;
    
    // Allocate memory for environment variables
    envp = (char **)malloc(sizeof(char *) * (NUM_ENV_VARS + 1));
    if (!envp) {
        perror("malloc failed");
        exit(1);
    }
    
    // Create a large number of environment variables
    // This should result in a large auxiliary vector
    for (i = 0; i < NUM_ENV_VARS; i++) {
        envp[i] = malloc(100);
        if (!envp[i]) {
            perror("malloc failed");
            exit(1);
        }
        snprintf(envp[i], 100, "VAR%d=value%d", i, i);
    }
    envp[NUM_ENV_VARS] = NULL;
    
    pid = fork();
    if (pid == 0) {
        // Child process - attempt to exec with large environment
        // This should trigger the auxvec size calculation in the kernel
        execve("/bin/true", (char *[]){"true", NULL}, envp);
        
        // If we reach here, execve failed
        perror("execve failed");
        exit(1);
    } else if (pid > 0) {
        // Parent process - wait for child
        int status;
        waitpid(pid, &status, 0);
        
        // Clean up
        for (i = 0; i < NUM_ENV_VARS; i++) {
            free(envp[i]);
        }
        free(envp);
        
        printf("PoC completed\n");
    } else {
        perror("fork failed");
        exit(1);
    }
    
    return 0;
}

```

### Environment

- Official docker asterinas/asterinas:0.17.0-20251217
- Asterinas version: main [Add unit tests for DMA `alloc` and `alloc_uninit()` · asterinas/asterinas@8ace769](https://github.com/asterinas/asterinas/commit/8ace76978eadd74dd2952866c269ac88b41c577b)

### Logs

```
./ex1004___asterinas__kernel__src__process__process_vm__init_stack__mod.rs_308_28
[   119.808] ERROR: Uncaught panic:
        attempt to subtract with overflow
        at /root/asterinas/kernel/src/process/process_vm/init_stack/mod.rs:381
        on CPU 0 by thread Some(Thread { task: (Weak), data: Any { .. }, is_exited: false, cpu_affinity: AtomicIdSet { bits: [1], phantom: PhantomData<ostd::cpu::id::CpuId> }, sched_attr: SchedAttr { policy: SchedPolicyState { kind: AtomicSchedPolicyKind(2), policy: UnsafeCell { .. } }, last_cpu: AtomicCpuId(0), real_time: RealTimeAttr { prio: 99, time_slice: 31490680 }, fair: FairAttr { weight: 1024, pending_weight: 0, vruntime: 57075327464 } } })
Printing stack trace:
   1: fn 0xffffffff88970160 - pc 0xffffffff889701ab / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce6f80;

   2: fn 0xffffffff882187c0 - pc 0xffffffff88218fbd / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce6fa0;

   3: fn 0xffffffff8821a780 - pc 0xffffffff8821a78a / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce7370;

   4: fn 0xffffffff8804b000 - pc 0xffffffff8804b00a / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce7380;

   5: fn 0xffffffff88ae4210 - pc 0xffffffff88ae4256 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce7390;

   6: fn 0xffffffff88aee680 - pc 0xffffffff88aee6a5 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce73d0;

   7: fn 0xffffffff8827ccd0 - pc 0xffffffff8827cdac / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce73e0;

   8: fn 0xffffffff8827d240 - pc 0xffffffff8827d293 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce7470;

   9: fn 0xffffffff8827d5c0 - pc 0xffffffff8827d670 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce74c0;

  10: fn 0xffffffff8827e190 - pc 0xffffffff8827e1f9 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce75a0;

  11: fn 0xffffffff8827f120 - pc 0xffffffff8827f728 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce7aa0;

  12: fn 0xffffffff882f29f0 - pc 0xffffffff882f2abb / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce8180;

  13: fn 0xffffffff88214af0 - pc 0xffffffff88214f6e / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce8220;

  14: fn 0xffffffff8858ed00 - pc 0xffffffff8858edd2 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce8700;

  15: fn 0xffffffff886d6330 - pc 0xffffffff886d6a0c / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce8860;

  16: fn 0xffffffff8865a570 - pc 0xffffffff8865a6f0 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce8f80;

  17: fn 0xffffffff885f89f0 - pc 0xffffffff88608280 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffce90f0;

  18: fn 0xffffffff885f6a50 - pc 0xffffffff885f6aea / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfc4f0;

  19: fn 0xffffffff8807e920 - pc 0xffffffff8807f1d1 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfc650;

  20: fn 0xffffffff8807f390 - pc 0xffffffff8807f3ee / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcb40;

  21: fn 0xffffffff886617f0 - pc 0xffffffff8866181e / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfccf0;

  22: fn 0xffffffff88661720 - pc 0xffffffff88661766 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcd20;

  23: fn 0xffffffff885c94f0 - pc 0xffffffff885c951c / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcd90;

  24: fn 0xffffffff8813b520 - pc 0xffffffff8813b55f / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcdd0;

  25: fn 0xffffffff8807f350 - pc 0xffffffff8807f377 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcf40;

  26: fn 0xffffffff8805f290 - pc 0xffffffff8805f2ad / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcf80;

  27: fn 0xffffffff88945e40 - pc 0xffffffff88945e65 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcfa0;

  28: fn 0xffffffff889631a0 - pc 0xffffffff88963224 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfcfd0;

  29: fn 0xffffffff8896df2d - pc 0xffffffff8896df36 / registers:

     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfd000;


     rax 0xffffdfffffce6e98; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffcfd008;

OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) ostd::panic::print_stack_trace
          at /root/asterinas/ostd/src/panic.rs:113
(  2) aster_kernel::thread::oops::panic_handler
          at /root/asterinas/kernel/src/thread/oops.rs:130
(  3) __ostd_panic_handler
          at ??:?
(  4) __rustc::rust_begin_unwind
          at 5hffciyz3d33t9gfe7c8dzufi:?
(  5) core::panicking::panic_fmt
          at ??:?
(  6) core::panicking::panic_const::panic_const_sub_overflow
          at ??:?
(  7) <aster_kernel::process::process_vm::init_stack::InitStackWriter>::write_bytes
          at 5xxl6ug2pujcy0lkgfpx9wyt0:?
(  8) <aster_kernel::process::process_vm::init_stack::InitStackWriter>::write_cstring
          at /root/asterinas/kernel/src/process/process_vm/init_stack/mod.rs:372
(  9) <aster_kernel::process::process_vm::init_stack::InitStackWriter>::write_envp_strings
          at /root/asterinas/kernel/src/process/process_vm/init_stack/mod.rs:285
( 10) <aster_kernel::process::process_vm::init_stack::InitStackWriter>::write
          at /root/asterinas/kernel/src/process/process_vm/init_stack/mod.rs:253
( 11) <aster_kernel::process::process_vm::init_stack::InitStack>::map_and_write
          at /root/asterinas/kernel/src/process/process_vm/init_stack/mod.rs:196
( 12) <aster_kernel::process::process_vm::ProcessVm>::map_and_write_init_stack
          at /root/asterinas/kernel/src/process/process_vm/mod.rs:124
( 13) aster_kernel::process::program_loader::elf::load_elf::load_elf_to_vmar
          at /root/asterinas/kernel/src/process/program_loader/elf/load_elf.rs:70
( 14) <aster_kernel::process::program_loader::ProgramToLoad>::load_to_vmar
          at /root/asterinas/kernel/src/process/program_loader/mod.rs:91
( 15) aster_kernel::process::execve::do_execve
          at /root/asterinas/kernel/src/process/execve.rs:58
( 16) aster_kernel::syscall::execve::sys_execve
          at /root/asterinas/kernel/src/syscall/execve.rs:27
( 17) aster_kernel::syscall::arch::syscall_dispatch
          at /root/asterinas/kernel/src/syscall/mod.rs:220
( 18) aster_kernel::syscall::handle_syscall
          at /root/asterinas/kernel/src/syscall/mod.rs:374
( 19) aster_kernel::thread::task::create_new_user_task::{closure#0}
          at /root/asterinas/kernel/src/thread/task.rs:95
( 20) aster_kernel::thread::task::create_new_user_task::{closure#1}
          at /root/asterinas/kernel/src/thread/task.rs:126
( 21) unwinding::panicking::catch_unwind::do_call::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:56
( 22) unwinding::panicking::catch_unwind::<unwinding::panic::RustPanic, (), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:42
( 23) unwinding::panic::catch_unwind::<(), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panic.rs:87
( 24) aster_kernel::thread::oops::catch_panics_as_oops::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/asterinas/kernel/src/thread/oops.rs:55
( 25) aster_kernel::thread::task::create_new_user_task::{closure#2}
          at /root/asterinas/kernel/src/thread/task.rs:131
( 26) <aster_kernel::thread::task::create_new_user_task::{closure#2} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 27) <alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output = ()> + core::marker::Send> as core::ops::function::FnOnce<()>>::call_once
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2206
( 28) kernel_task_entry
          at /root/asterinas/ostd/src/task/mod.rs:196
( 29) kernel_task_entry_wrapper
          at ??:?
make: *** [Makefile:309: run_kernel] Error 1
```

--- a/kernel/src/process/process_vm/init_stack/mod.rs
+++ b/kernel/src/process/process_vm/init_stack/mod.rs
@@ -18,7 +18,6 @@ use core::{
     sync::atomic::{AtomicUsize, Ordering},
 };
 
-use align_ext::AlignExt;
 use ostd::mm::{MAX_USERSPACE_VADDR, VmIo};
 
 use self::aux_vec::{AuxKey, AuxVec};
@@ -359,10 +358,9 @@ impl InitStackWriter<'_> {
     /// Writes u64 to the stack.
     /// Returns the writing address
     fn write_u64(&self, val: u64) -> Result<u64> {
-        let start_address = (self.pos() - 8).align_down(8);
-        self.pos.store(start_address, Ordering::Relaxed);
-        self.vmo.write_val(start_address - self.map_addr, &val)?;
-        Ok(self.pos() as u64)
+        let new_pos = self.reserve_pos(size_of::<u64>(), align_of::<u64>())?;
+        self.vmo.write_val(new_pos - self.map_addr, &val)?;
+        Ok(new_pos as u64)
     }
 
     /// Writes a CString including the ending null byte to the stack.
@@ -372,14 +370,23 @@ impl InitStackWriter<'_> {
         self.write_bytes(bytes)
     }
 
-    /// Writes u64 to the stack.
+    /// Writes bytes to the stack.
     /// Returns the writing address.
     fn write_bytes(&self, bytes: &[u8]) -> Result<u64> {
-        let len = bytes.len();
-        self.pos.fetch_sub(len, Ordering::Relaxed);
-        let pos = self.pos();
-        self.vmo.write_bytes(pos - self.map_addr, bytes)?;
-        Ok(pos as u64)
+        let new_pos = self.reserve_pos(bytes.len(), align_of::<u8>())?;
+        self.vmo.write_bytes(new_pos - self.map_addr, bytes)?;
+        Ok(new_pos as u64)
+    }
+
+    fn reserve_pos(&self, size: usize, align: usize) -> Result<Vaddr> {
+        if let Some(pos) = self.pos().checked_sub(size)
+            && let new_pos = pos & !(align - 1)
+            && new_pos >= self.map_addr
+        {
+            self.pos.store(new_pos, Ordering::Relaxed);
+            return Ok(new_pos);
+        }
+        return_errno_with_message!(Errno::E2BIG, "Init stack overflow");
     }
 
     fn pos(&self) -> Vaddr {

